File: app/src/main/java/com/arnaudpiroelle/manga/service/MangaDownloadManager.java
package com.arnaudpiroelle.manga.service;

import android.os.Environment;
import android.support.annotation.VisibleForTesting;
import android.util.Log;

import com.arnaudpiroelle.manga.core.provider.MangaProvider;
import com.arnaudpiroelle.manga.core.provider.ProviderRegistry;
import com.arnaudpiroelle.manga.core.utils.FileHelper;
import com.arnaudpiroelle.manga.core.utils.HttpUtils;
import com.arnaudpiroelle.manga.model.Chapter;
import com.arnaudpiroelle.manga.model.Manga;
import com.arnaudpiroelle.manga.model.Page;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import rx.Observable;
import rx.schedulers.Schedulers;

public class MangaDownloadManager {

    private MangaDownloaderCallback callback;
    private ProviderRegistry providerRegistry;
    private FileHelper fileHelper;

    public MangaDownloadManager(MangaDownloaderCallback callback, ProviderRegistry providerRegistry, FileHelper fileHelper) {
        this.callback = callback;
        this.providerRegistry = providerRegistry;
        this.fileHelper = fileHelper;
    }

    public void startDownload(List<Manga> mangas) {
        getValidMangas(mangas)
                .subscribeOn(Schedulers.io())
                .doOnNext(this::onMangaDownload)
                .doOnCompleted(this::onAllMangasDownloaded)
                .doOnError(this::onError)
                .flatMap(this::downloadManga)
                .subscribe();

    }

    @VisibleForTesting Observable<Manga> getValidMangas(List<Manga> mangas){
        return Observable.from(mangas)
                .filter(this::hasValidProvider);
    }

    @VisibleForTesting Observable<Chapter> getNewChapters(Manga manga){
        MangaProvider mangaProvider = providerRegistry.get(manga.getProvider());

        List<Chapter> chapters = mangaProvider.findChapters(manga);
        manga.setChapters(chapters);

        return Observable.from(chapters)
                .skip(alreadyDownloadedChapters(manga));
    }

    @VisibleForTesting Observable<Page> getPages(Manga manga, Chapter chapter){
        MangaProvider mangaProvider = providerRegistry.get(manga.getProvider());

        List<Page> pages = mangaProvider.findPages(chapter);
        chapter.setPages(pages);

        return Observable.from(pages)
                .doOnNext(page -> onPageDownload(manga, chapter, page));
    }

    private Observable<?> downloadManga(final Manga manga) {
        return getNewChapters(manga)
                .doOnNext(this::onChapterDownload)
                .doOnCompleted(() -> onAllChaptersDownloaded(manga))
                .doOnError(this::onError)
                .flatMap(chapter -> downloadChapter(manga, chapter));
    }

    private Observable<?> downloadChapter(final Manga manga, final Chapter chapter) {
        return getPages(manga, chapter)
                .doOnNext(page -> onPageDownload(manga, chapter, page))
                .doOnCompleted(() -> onAllPagesDownloaded(manga, chapter))
                .doOnError(this::onError)
                .flatMap(page -> downloadPage(manga, chapter, page));
    }

    private Observable<File> downloadPage(final Manga manga, final Chapter chapter, final Page page) {
        return Observable.<File>create(subscriber -> {
            InputStream inputStream = providerRegistry.get(manga.getProvider()).findPage(page);

            File pageFile = fileHelper.getPageFile(manga, chapter, page);

            try {
                pageFile.createNewFile();
                HttpUtils.writeFile(inputStream, pageFile);
            } catch (IOException e) {
                Log.e("DownloadService", "Error", e);
            }

            subscriber.onNext(pageFile);
            subscriber.onCompleted();
        }).doOnError(this::onError);
    }

    private void onError(Throwable throwable) {
        callback.onDownloadError(throwable);
    }

    @VisibleForTesting Boolean hasValidProvider(Manga manga) {
        return providerRegistry.get(manga.getProvider()) != null;
    }

    @VisibleForTesting int alreadyDownloadedChapters(Manga manga) {
        if (manga.getLastChapter() == null || manga.getLastChapter().isEmpty()) {
            return manga.getChapters().size() - 1;
        } else {
            if ("all".equals(manga.getLastChapter())) {
                return 0;
            } else {
                for (int i = 0; i < manga.getChapters().size(); i++) {
                    if (manga.getChapters().get(i).getChapterNumber().equals(manga.getLastChapter())) {
                        return i + 1;
                    }
                }
            }
        }

        return 0;
    }

    private void onAllMangasDownloaded() {
        Log.i("NewMangaDownloader", "All mangas downloaded");
        callback.onDownloadCompleted();
    }

    private void onMangaDownload(Manga manga) {
        Log.i("NewMangaDownloader", "Manga " + manga.getName() + " download");
    }

    private void onAllChaptersDownloaded(Manga manga) {
        Log.i("NewMangaDownloader", "All chapters downloaded");
        callback.onCompleteManga(manga);
    }

    private void onChapterDownload(Chapter chapter) {
        Log.i("NewMangaDownloader", "Chapter " + chapter.getChapterNumber() + " download");
    }

    private void onAllPagesDownloaded(Manga manga, Chapter chapter) {
        Log.i("NewMangaDownloader", "All pages downloaded");
        callback.onCompleteChapter(manga, chapter);
    }

    private void onPageDownload(Manga manga, Chapter chapter, Page page) {
        Log.i("NewMangaDownloader", "Page " + page.getPageNumber() + "/" + chapter.getPages().size() + " download");
        callback.onCompletePage(manga, chapter, page);
    }

    public void zipChapter(final Manga manga, final Chapter chapter) {

        Observable.create(subscriber -> {
            File mangaFolder = fileHelper.getMangaFolder(manga);

            File zipFile = new File(String.format("%s/%s - %s.cbz",
                    mangaFolder.getAbsoluteFile(),
                    manga.getName(),
                    chapter.getChapterNumber()));

            try (FileOutputStream dest = new FileOutputStream(zipFile);
                 ZipOutputStream out = new ZipOutputStream(new BufferedOutputStream(dest))) {

                zipFile.createNewFile();
                File chapterFolder = fileHelper.getChapterFolder(manga, chapter);

                int buffer = 1024;
                byte data[] = new byte[buffer];

                for (File file : chapterFolder.listFiles()) {
                    try (FileInputStream fi = new FileInputStream(file);
                         BufferedInputStream origin = new BufferedInputStream(fi, buffer)) {

                        ZipEntry entry = new ZipEntry(file.getName());
                        out.putNextEntry(entry);

                        int count;
                        while ((count = origin.read(data, 0, buffer)) != -1) {
                            out.write(data, 0, count);
                        }
                    }
                }

                for (File file : chapterFolder.listFiles()) {
                    file.delete();
                }

                chapterFolder.delete();

            } catch (IOException e) {
                Log.e("MangaDownloadManager", "Error when zip chapter");

                if (zipFile.exists()) {
                    zipFile.delete();
                }
            } finally {
                subscriber.onCompleted();
            }
        })
                .subscribeOn(Schedulers.io())
                .subscribe();

    }

    public interface MangaDownloaderCallback {

        void onDownloadError(Throwable throwable);

        void onDownloadCompleted();

        void onCompleteManga(Manga manga);

        void onCompleteChapter(Manga manga, Chapter chapter);

        void onCompletePage(Manga manga, Chapter chapter, Page page);
    }
}

File: app/src/main/kotlin/com/arnaudpiroelle/manga/service/MangaDownloadManager.kt
package com.arnaudpiroelle.manga.service

import android.support.annotation.VisibleForTesting
import android.util.Log
import com.arnaudpiroelle.manga.core.provider.MangaProvider
import com.arnaudpiroelle.manga.core.provider.ProviderRegistry
import com.arnaudpiroelle.manga.core.utils.FileHelper
import com.arnaudpiroelle.manga.core.utils.HttpUtils
import com.arnaudpiroelle.manga.model.Chapter
import com.arnaudpiroelle.manga.model.Manga
import com.arnaudpiroelle.manga.model.Page
import rx.Observable
import rx.Observable.from
import rx.schedulers.Schedulers
import java.io.*
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

class MangaDownloadManager(private val callback: MangaDownloadManager.MangaDownloaderCallback, private val providerRegistry: ProviderRegistry, private val fileHelper: FileHelper) {

    fun startDownload(provider: MangaProvider, mangas: List<Manga>) {
        from(mangas)
                .subscribeOn(Schedulers.io())
                .forEach(
                        {
                            onMangaDownload(it)
                            downloadManga(provider, it)
                        },
                        { onError(it) },
                        { onAllMangasDownloaded() }
                )
    }

    private fun downloadManga(provider: MangaProvider, manga: Manga) {
        getChapters(provider, manga).forEach (
                {
                    onChapterDownload(it)
                    downloadChapter(provider, manga, it)
                },
                { onError(it) },
                { onAllChaptersDownloaded(manga) }
        )
    }

    @VisibleForTesting fun getChapters(provider: MangaProvider, manga: Manga): Observable<Chapter> {
        val chapters = provider.findChapters(manga)
        manga.chapters = chapters

        return from(chapters)
                .skip(alreadyDownloadedChapters(manga))
    }

    private fun downloadChapter(provider: MangaProvider, manga: Manga, chapter: Chapter) {
        getPages(provider, manga, chapter)
                .forEach (
                        {
                            onPageDownload(manga, chapter, it)
                            downloadPage(provider, manga, chapter, it)
                        },
                        { onError(it) },
                        { onAllPagesDownloaded(manga, chapter) }
                )
    }

    fun getPages(provider: MangaProvider, manga: Manga, chapter: Chapter): Observable<Page> {
        val pages = provider.findPages(chapter)
        chapter.pages = pages

        return from(pages)
    }

    private fun downloadPage(provider: MangaProvider, manga: Manga, chapter: Chapter, page: Page) {
        val inputStream = provider.findPage(page)

        fileHelper.getPageFile(manga, chapter, page).apply {
            createNewFile()
            HttpUtils.writeFile(inputStream, this)
            postProcess(provider, this)

        }

    }

    private fun onError(throwable: Throwable) {
        callback.onDownloadError(throwable)
    }

    @VisibleForTesting fun alreadyDownloadedChapters(manga: Manga): Int {
        if (manga.lastChapter == null || manga.lastChapter?.isEmpty()!!) {
            return manga.chapters?.size!! - 1
        } else {
            if ("all" == manga.lastChapter) {
                return 0
            } else {
                for (i in 0..manga.chapters?.size!! - 1) {
                    if (manga.chapters!!.get(i).chapterNumber == manga.lastChapter) {
                        return i + 1
                    }
                }
            }
        }

        return 0
    }

    private fun onAllMangasDownloaded() {
        Log.i("NewMangaDownloader", "All mangas downloaded")
        callback.onDownloadCompleted()
    }

    private fun onMangaDownload(manga: Manga) {
        Log.i("NewMangaDownloader", "Manga " + manga.name + " download")
    }

    private fun onAllChaptersDownloaded(manga: Manga) {
        Log.i("NewMangaDownloader", "All chapters downloaded")
        callback.onCompleteManga(manga)
    }

    private fun onChapterDownload(chapter: Chapter) {
        Log.i("NewMangaDownloader", "Chapter " + chapter.chapterNumber + " download")
    }

    private fun onAllPagesDownloaded(manga: Manga, chapter: Chapter) {
        Log.i("NewMangaDownloader", "All pages downloaded")
        callback.onCompleteChapter(manga, chapter)
    }

    private fun onPageDownload(manga: Manga, chapter: Chapter, page: Page) {
        Log.i("NewMangaDownloader", "Page " + page.pageNumber + "/" + chapter.pages?.size + " download")
        callback.onCompletePage(manga, chapter, page)
    }

    fun zipChapter(manga: Manga, chapter: Chapter) {

        Observable.create<Any> { subscriber ->
            val mangaFolder = fileHelper.getMangaFolder(manga)

            val zipFile = File("%s/%s - %s.cbz".format(mangaFolder.absoluteFile, manga.name, chapter.chapterNumber))

            try {
                FileOutputStream(zipFile).use { dest ->
                    ZipOutputStream(BufferedOutputStream(dest)).use { out ->

                        zipFile.createNewFile()
                        val chapterFolder = fileHelper.getChapterFolder(manga, chapter)

                        val buffer = 1024
                        val data = ByteArray(buffer)

                        for (file in chapterFolder.listFiles()) {
                            FileInputStream(file).use { fi ->
                                BufferedInputStream(fi, buffer).use { origin ->

                                    val entry = ZipEntry(file.name)
                                    out.putNextEntry(entry)

                                    var count = origin.read(data, 0, buffer)
                                    while (count != -1) {
                                        out.write(data, 0, count)
                                        count = origin.read(data, 0, buffer)
                                    }
                                }
                            }
                        }

                        for (file in chapterFolder.listFiles()) {
                            file.delete()
                        }

                        chapterFolder.delete()

                    }
                }
            } catch (e: IOException) {
                Log.e("MangaDownloadManager", "Error when zip chapter")

                if (zipFile.exists()) {
                    zipFile.delete()
                }
            } finally {
                subscriber.onCompleted()
            }
        }.subscribeOn(Schedulers.io()).subscribe()

    }

    fun postProcess(provider: MangaProvider, file: File) {
        Log.i("PostProcess", "post process : ${file.absolutePath}")
        provider.postProcess(file)
    }

    interface MangaDownloaderCallback {

        fun onDownloadError(throwable: Throwable)

        fun onDownloadCompleted()

        fun onCompleteManga(manga: Manga)

        fun onCompleteChapter(manga: Manga, chapter: Chapter)

        fun onCompletePage(manga: Manga, chapter: Chapter, page: Page)
    }
}
